src/libostree/ostree-repo.c

```diff
@@ -405,26 +412,40 @@ pop_repo_lock (OstreeRepo  *self,
   GHashTable *lock_table = g_private_get (&repo_lock_table);
   g_return_val_if_fail (lock_table != NULL, FALSE);
 
-  OstreeRepoLock *lock = g_hash_table_lookup (lock_table, self);
-  g_return_val_if_fail (lock != NULL, FALSE);
-  g_return_val_if_fail (lock->fd != -1, FALSE);
+  GPtrArray *stack = g_hash_table_lookup (lock_table, self);
+  g_return_val_if_fail (stack != NULL, FALSE);
+
+  g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&self->lock.mutex);
+  g_return_val_if_fail (self->lock.fd != -1, FALSE);
 
   OstreeRepoLockInfo info;
-  repo_lock_info (lock, &info);
+  repo_lock_info (self, &info);
   g_return_val_if_fail (info.len > 0, FALSE);
 
   g_debug ("Pop lock: state=%s, depth=%u", info.name, info.len);
+  int state_to_drop = GPOINTER_TO_INT (stack->pdata[stack->len - 1]);
   if (info.len > 1)
     {
-      int next_state = GPOINTER_TO_INT (g_queue_peek_nth (&lock->stack, 1));
+      int next_state;
+
+      if (state_to_drop == LOCK_EX)
+        {
+          g_return_val_if_fail (self->lock.exclusive > 0, FALSE);
```

Comment:
In ostree code I tend to prefer assertions over `g_return_val_if_fail` because we're maintaining the root filesystem and crashing is better than corruption.  
i.e.

```suggestion
          g_assert_cmpint (self->lock.exclusive, >, 0);
```

(Personally I'd say even for desktop apps this is better but that battle feels kind of lost)

Reply:
I'm fine with that. I had been using `g_return*` rather than `g_assert*` under the idea that there's less chance they would be disabled, but I guess it's possible either `G_DISABLE_ASSERT` or `G_DISABLE_CHECKS` could be defined.

Reply:
I changed all these checks to use `g_assert*` or `g_error`.